mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m

```diff
@@ -0,0 +1,211 @@
+/*
+ * Keyman is copyright (C) SIL International. MIT License.
+ *
+ * KMSettingsRepository.h
+ * Keyman
+ *
+ * Created by Shawn Schantz on 2024-07-29.
+ *
+ * Singleton object for reading and writing Keyman application settings.
+ * Serves as an abstraction to StandardUserDefaults which is currently used to store application settings.
+ */
+
+#import "KMSettingsRepository.h"
+#import "KMLogs.h"
+
+NSString *const kActiveKeyboardsKey = @"KMActiveKeyboardsKey";
+NSString *const kSelectedKeyboardKey = @"KMSelectedKeyboardKey";
+NSString *const kPersistedOptionsKey = @"KMPersistedOptionsKey";
+
+NSString *const kObsoletePathComponent = @"/Documents/";
+NSString *const kNewPathComponent = @"/Library/Application Support/keyman.inputmethod.Keyman/";
+
+/**
+ * Store the version number of the data model in the UserDefaults with this key.
+ * The first version, 1, is defined to indicate that we are storing the data/keyboards in the Library
+ * directory instead of in the Documents directory.
+ */
+NSString *const kDataModelVersion = @"KMDataModelVersion";
+NSInteger const kVersionStoreDataInLibraryDirectory = 1;
+NSInteger const kCurrentDataModelVersionNumber = kVersionStoreDataInLibraryDirectory;
+
+@implementation KMSettingsRepository
+
++ (KMSettingsRepository *)shared
+{
+  static KMSettingsRepository *shared = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    shared = [[KMSettingsRepository alloc] init];
+  });
+  return shared;
+}
+
+- (void)setDataModelVersionIfNecessary {
+  if (![self dataModelWithKeyboardsInLibrary]) {
+    [[NSUserDefaults standardUserDefaults] setInteger:kVersionStoreDataInLibraryDirectory forKey:kDataModelVersion];
+  }
+}
+
+/**
+ * If the selectedKeyboard has not been set, then the settings have not been saved in the UserDefaults.
+ * If this method is called after applicationDidFinishLaunching, then it will always return true.
+ * If called from awakeFromNib, then it will return false when running for the first time.
+ */
+- (BOOL)settingsExist {
+  return [[NSUserDefaults standardUserDefaults] objectForKey:kSelectedKeyboardKey] != nil;
+}
+
+/**
+ * For the first numbered version of the data model, the app stores the keyboards under the ~/Library directory
+ * For versions before version 1, the keyboards were stored under the ~/Documents directory.
+ */
+- (BOOL)dataModelWithKeyboardsInLibrary {
+  // NSUserDefaults returns zero if the key does not exist
```

Comment:
```suggestion
  // [NSUserDefaults integerForKey] returns zero if the key does not exist
```

Reply:
Created issue #12294 to document the need for better error handling in case the data directory cannot be located or is unreadable. The current code does not gracefully handle a nil returned from the OS when trying to access '~/Library/Application Support'